### PR TITLE
Implement the `Add`/`Div`/`Mul`/`Sub` blocks

### PR DESCRIPTION
**Pull request by oleksandrvoyager**

### CODE REVIEW

File: lib/protoflow-blocks/src/blocks/math/add.rs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    fn execute(&mut self, _runtime: &dyn BlockRuntime) -> BlockResult {
+        let mut result = None;
+        while let Some(input) = self.input.recv()? {
+            let res = match result {
+                None => input,
+                Some(current_result) => current_result + input,
+            };
+
+            result = Some(res);
+            self.output.send(&res)?;
```
Comment (SamuelSarle) [inline lib/protoflow-blocks/src/blocks/math/add.rs:80]: Did you intend to send a value on each iteration?

File: lib/protoflow-blocks/src/blocks/math/div.rs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+                    if input == 0.0 {
+                        return Err(BlockError::Other("Division by zero".into()));
+                    }
+
+                    current_result / input
+                }
+            };
+
+            result = Some(res);
+            self.output.send(&res)?;
```
Comment (SamuelSarle) [inline lib/protoflow-blocks/src/blocks/math/div.rs:86]: Did you intend to send a value on each iteration?

File: lib/protoflow-blocks/src/blocks/math/mul.rs [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    fn execute(&mut self, _runtime: &dyn BlockRuntime) -> BlockResult {
+        let mut product = None;
+        while let Some(input) = self.input.recv()? {
+            let res = match product {
+                None => input,
+                Some(current_product) => current_product * input,
+            };
+
+            product = Some(res);
+            self.output.send(&res)?;
```
Comment (SamuelSarle) [inline lib/protoflow-blocks/src/blocks/math/mul.rs:80]: Did you intend to send a value on each iteration?

File: lib/protoflow-blocks/src/blocks/math/sub.rs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    fn execute(&mut self, _runtime: &dyn BlockRuntime) -> BlockResult {
+        let mut result = None;
+        while let Some(input) = self.input.recv()? {
+            let res = match result {
+                None => input,
+                Some(current_result) => current_result - input,
+            };
+
+            result = Some(res);
+            self.output.send(&res)?;
```
Comment (SamuelSarle) [inline lib/protoflow-blocks/src/blocks/math/sub.rs:80]: Did you intend to send a value on each iteration?

**Review by SamuelSarle — CHANGES_REQUESTED** [review]
LGTM otherwise but I suggest to change the `while self.input.recv { ..., self.output.send(&res) }` to `while self.input.recv { ... } self.output.send(&result.unwrap_or(<default-value>))` with `<default-value>` being `0.0` for `Add` and `Sub` blocks, and `1.0` for `Mul` and `Div` blocks.
Reasoning for the defaults comes from the identity elements of [addition](https://en.wikipedia.org/wiki/Addition#Identity_element) or [multiplication](https://en.wikipedia.org/wiki/Product_(mathematics)#Empty_product), e.g. adding/subtracting `0` does nothing and multiplying/diving by `1` does nothing and thus they're used as the values when the input set is empty, adding no numbers together equals 0 and [multiplying no numbers](https://en.wikipedia.org/wiki/Empty_product) together equals 1.

Also noting that as a future improvement, it's quite easy to switch these implementations to support any numerical types (f32,f64,u32,u64, etc.) by defining the blocks with generics as such:

```rust
pub struct Add<T: Message + core::ops::Add<Output = T> + num_traits::Zero = f64> { /* ... */ }
pub struct Sub<T: Message + core::ops::Sub<Output = T> + num_traits::Zero = f64> { /* ... */ }
pub struct Mul<T: Message + core::ops::Mul<Output = T> + num_traits::One = f64> { /* ... */ }
pub struct Div<T: Message + core::ops::Div<Output = T> + PartialEq + num_traits::One + num_traits::Zero = f64> { /* ... */ }
```

In summary what those generics are saying is, taking for example the `Add` block:
1. Anything that implements `Message` (so that we can send it over a port)
2. And implements [`core::ops::Add`](https://doc.rust-lang.org/nightly/core/ops/trait.Add.html) (so that we can call `x + y` on the types)
3. And implements `num_traits::Zero` (so that we can call `T::zero()` for the identity element as the starting value for the summing)
4. `= f64`, by default use `f64` if nothing else was provided.

Let me know if you want the patch for those changes for this PR, otherwise I'll open another PR after this is merged.

File: lib/protoflow-blocks/src/blocks/math/div.rs (hunk 2 above)
Comment (AlexSanches1) [reply lib/protoflow-blocks/src/blocks/math/div.rs:86]: yes

**Review by SamuelSarle — APPROVED** [review]

**Comment by artob** [comment]
@race-of-sloths invite

**Comment by race-of-sloths** [comment]
@artob Thank you for calling!
    
@AlexSanches1 Thank you for the contribution! Join [Race of Sloths](https://race-of-sloths.com) by simply mentioning me in your comment/PRs description and start collecting Sloth Points through contributions to open source projects.

<details><summary>What is the Race of Sloths</summary>
<p>
Race of Sloths is a friendly competition where you can participate in challenges and compete with other open-source contributors within your normal workflow

For contributors:

- Tag @race-of-sloths inside your pull requests
- Wait for the maintainer to review and score your pull request
- Check out your position in the [Leaderboard](https://race-of-sloths.com/leaderboard)
- Keep weekly and monthly streaks to reach higher positions
- Boast your contributions with a dynamic picture of your [Profile](https://race-of-sloths.com/profile/AlexSanches1)

For maintainers:
- Score pull requests that participate in the Race of Sloths and receive a reward
- Engage contributors with fair scoring and fast responses so they keep their streaks
- Promote the Race to the point where the Race starts promoting you
- Grow the community of your contributors

Feel free to check [our website](https://race-of-sloths.com) for additional details!
</p>
</details> 

<details>
<summary>Bot commands</summary>

- For contributors
  - **Include a PR:** `@race-of-sloths include` to enter the Race with your PR
- For maintainers:
  - **Invite contributor** `@race-of-sloths invite` to invite the contributor to participate in a race or include it, if it's already a runner.
  - **Assign points:** `@race-of-sloths score [1/2/3/5/8/13]` to award points based on your assessment.
  - **Reject this PR:** `@race-of-sloths exclude` to send this PR back to the drawing board.
  - **Exclude repo:** `@race-of-sloths pause` to stop bot activity in this repo until `@race-of-sloths unpause` command is called

</details>

**Comment by artob** [comment]
> [!IMPORTANT]
> **Protoflow has been discontinued** and archived in favor of its successors [Async-Flow](https://github.com/artob/async-flow) and [Flows.rs](https://github.com/artob/flows.rs), which represent the future of flow-based programming in Rust and offer a substantially better developer experience as well as simply higher development velocity. We aim to eventually port the functionality in this pull request over to Flows.rs; if you'd like to contribute to make that happen sooner, please raise an issue and/or pull request against that repository.
